Build with SYCL fails using intel/llvm compiler

With:
* oneDNN [v3.5.3](https://github.com/oneapi-src/oneDNN/releases/tag/v3.5.3) or 817732f46c70b2b7d6ff0c4cc668536347448636
* https://github.com/intel/llvm/commit/3b91b0bf4498c031a67abb6310addaad83a99438
* https://github.com/oneapi-src/oneTBB/commit/ce60d6b7f97f45cc6b2238fe0ca843c1e250dcc1

Build fails if configured in the following way (that's pytorch style build, see https://github.com/pytorch/pytorch/blob/a9d34138dfd5c049305487fc9f1b8bae0cadc98e/cmake/Modules/FindMKLDNN.cmake#L48):
```
cmake \
  -DCMAKE_C_COMPILER=icx \
  -DCMAKE_CXX_COMPILER=icpx \
  -DDNNL_GPU_RUNTIME=SYCL \
  -DDNNL_CPU_RUNTIME=THREADPOOL \
  -DDNNL_BUILD_TESTS=OFF \
  -DDNNL_BUILD_EXAMPLES=OFF \
  -DONEDNN_BUILD_GRAPH=OFF \
  -DDNNL_LIBRARY_TYPE=STATIC \
  -DDNNL_DPCPP_HOST_COMPILER=g++ \
  ..
```
Note: icx/icpx are symbolic links to `clang-19` built in intel/llvm.

Few files fail with the similar error. For example:
```
In file included from <command-line>:
/tmp/sycl_engine_base-header-11d9b7.h:46:20: error: ‘KernelInfoData’ is not a class template
   46 | template <> struct KernelInfoData<'_', 'Z', 'T', 'S', 'N', '4', 's', 'y', 'c', 'l', '3', '_', 'V', '1', '6', 'd', 'e', 't', 'a', 'i', 'l', '1', '8', 'R', 'o', 'u', 'n', 'd', 'e', 'd', 'R', 'a', 'n',
 'g', 'e', 'K', 'e', 'r', 'n', 'e', 'l', 'I', 'N', 'S', '0', '_', '4', 'i', 't', 'e', 'm', 'I', 'L', 'i', '1', 'E', 'L', 'b', '1', 'E', 'E', 'E', 'L', 'i', '1', 'E', 'Z', 'N', 'S', '0', '_', '7', 'h', 'a',
'n', 'd', 'l', 'e', 'r', '1', '8', 'c', 'o', 'p', 'y', 'A', 'c', 'c', 'T', 'o', 'A', 'c', 'c', 'H', 'e', 'l', 'p', 'e', 'r', 'I', 'h', 'L', 'i', '1', 'E', 'L', 'N', 'S', '0', '_', '6', 'a', 'c', 'c', 'e', '
s', 's', '4', 'm', 'o', 'd', 'e', 'E', '1', '0', '2', '4', 'E', 'L', 'N', 'S', '7', '_', '6', 't', 'a', 'r', 'g', 'e', 't', 'E', '2', '0', '1', '4', 'E', 'h', 'L', 'i', '1', 'E', 'L', 'S', '8', '_', '1', '0
', '2', '5', 'E', 'L', 'S', '9', '_', '2', '0', '1', '4', 'E', 'L', 'N', 'S', '7', '_', '1', '1', 'p', 'l', 'a', 'c', 'e', 'h', 'o', 'l', 'd', 'e', 'r', 'E', '0', 'E', 'L', 'S', 'A', '_', '0', 'E', 'E', 'E'
, 'N', 'S', 't', '9', 'e', 'n', 'a', 'b', 'l', 'e', '_', 'i', 'f', 'I', 'X', 'a', 'a', 'g', 't', 'T', '0', '_', 'L', 'i', '0', 'E', 'g', 't', 'T', '4', '_', 'L', 'i', '0', 'E', 'E', 'b', 'E', '4', 't', 'y',
 'p', 'e', 'E', 'N', 'S', '0', '_', '8', 'a', 'c', 'c', 'e', 's', 's', 'o', 'r', 'I', 'T', '_', 'X', 'T', '0', '_', 'E', 'X', 'T', '1', '_', 'E', 'X', 'T', '2', '_', 'E', 'X', 'T', '7', '_', 'E', 'N', 'S',
'0', '_', '3', 'e', 'x', 't', '6', 'o', 'n', 'e', 'a', 'p', 'i', '2', '2', 'a', 'c', 'c', 'e', 's', 's', 'o', 'r', '_', 'p', 'r', 'o', 'p', 'e', 'r', 't', 'y', '_', 'l', 'i', 's', 't', 'I', 'J', 'E', 'E', '
E', 'E', 'E', 'N', 'S', 'E', '_', 'I', 'T', '3', '_', 'X', 'T', '4', '_', 'E', 'X', 'T', '5', '_', 'E', 'X', 'T', '6', '_', 'E', 'X', 'T', '8', '_', 'E', 'S', 'J', '_', 'E', 'E', 'E', 'U', 'l', 'N', 'S', '0
', '_', '2', 'i', 'd', 'I', 'L', 'i', '1', 'E', 'E', 'E', 'E', '_', 'E', 'E'> {
      |                    ^~~~~~~~~~~~~~
In file included from <command-line>:
/tmp/sycl_engine_base-header-11d9b7.h:46:1925: error: explicit specialization of non-template ‘sycl::_V1::detail::KernelInfoData’
   46 | template <> struct KernelInfoData<'_', 'Z', 'T', 'S', 'N', '4', 's', 'y', 'c', 'l', '3', '_', 'V', '1', '6', 'd', 'e', 't', 'a', 'i', 'l', '1', '8', 'R', 'o', 'u', 'n', 'd', 'e', 'd', 'R', 'a', 'n',
 'g', 'e', 'K', 'e', 'r', 'n', 'e', 'l', 'I', 'N', 'S', '0', '_', '4', 'i', 't', 'e', 'm', 'I', 'L', 'i', '1', 'E', 'L', 'b', '1', 'E', 'E', 'E', 'L', 'i', '1', 'E', 'Z', 'N', 'S', '0', '_', '7', 'h', 'a',
'n', 'd', 'l', 'e', 'r', '1', '8', 'c', 'o', 'p', 'y', 'A', 'c', 'c', 'T', 'o', 'A', 'c', 'c', 'H', 'e', 'l', 'p', 'e', 'r', 'I', 'h', 'L', 'i', '1', 'E', 'L', 'N', 'S', '0', '_', '6', 'a', 'c', 'c', 'e', '
s', 's', '4', 'm', 'o', 'd', 'e', 'E', '1', '0', '2', '4', 'E', 'L', 'N', 'S', '7', '_', '6', 't', 'a', 'r', 'g', 'e', 't', 'E', '2', '0', '1', '4', 'E', 'h', 'L', 'i', '1', 'E', 'L', 'S', '8', '_', '1', '0
', '2', '5', 'E', 'L', 'S', '9', '_', '2', '0', '1', '4', 'E', 'L', 'N', 'S', '7', '_', '1', '1', 'p', 'l', 'a', 'c', 'e', 'h', 'o', 'l', 'd', 'e', 'r', 'E', '0', 'E', 'L', 'S', 'A', '_', '0', 'E', 'E', 'E'
, 'N', 'S', 't', '9', 'e', 'n', 'a', 'b', 'l', 'e', '_', 'i', 'f', 'I', 'X', 'a', 'a', 'g', 't', 'T', '0', '_', 'L', 'i', '0', 'E', 'g', 't', 'T', '4', '_', 'L', 'i', '0', 'E', 'E', 'b', 'E', '4', 't', 'y',
 'p', 'e', 'E', 'N', 'S', '0', '_', '8', 'a', 'c', 'c', 'e', 's', 's', 'o', 'r', 'I', 'T', '_', 'X', 'T', '0', '_', 'E', 'X', 'T', '1', '_', 'E', 'X', 'T', '2', '_', 'E', 'X', 'T', '7', '_', 'E', 'N', 'S',
'0', '_', '3', 'e', 'x', 't', '6', 'o', 'n', 'e', 'a', 'p', 'i', '2', '2', 'a', 'c', 'c', 'e', 's', 's', 'o', 'r', '_', 'p', 'r', 'o', 'p', 'e', 'r', 't', 'y', '_', 'l', 'i', 's', 't', 'I', 'J', 'E', 'E', '
E', 'E', 'E', 'N', 'S', 'E', '_', 'I', 'T', '3', '_', 'X', 'T', '4', '_', 'E', 'X', 'T', '5', '_', 'E', 'X', 'T', '6', '_', 'E', 'X', 'T', '8', '_', 'E', 'S', 'J', '_', 'E', 'E', 'E', 'U', 'l', 'N', 'S', '0
', '_', '2', 'i', 'd', 'I', 'L', 'i', '1', 'E', 'E', 'E', 'E', '_', 'E', 'E'> {
      |
```

For individual file this issue can be reproduced with:
```
/home/gta/git/install/bin/icpx -DCL_TARGET_OPENCL_VERSION=300 -DDNNL_ENABLE_CPU_ISA_HINTS -DDNNL_ENABLE_ITT_TASKS -DDNNL_ENABLE_MAX_CPU_ISA -DDNNL_TEST_THREADPOOL_USE_STANDALONE -DDNNL_X64=1 -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -I/home/gta/git/oneDNN/_build/include -I/home/gta/git/oneDNN/include -I/home/gta/git/oneDNN/src -I/home/gta/git/oneDNN/src/sycl -fno-sycl-id-queries-fit-in-int -Wno-#pragma-messages -fsycl-host-compiler-options="-I/home/gta/git/oneDNN/_build/include -I/home/gta/git/oneDNN/include -DCL_TARGET_OPENCL_VERSION=300 -DDNNL_X64=1 -D__STDC_LIMIT_MACROS -D__STDC_CONSTANT_MACROS  -Wall -Wno-unknown-pragmas -fvisibility=internal -fvisibility-inlines-hidden -fPIC -Wformat -Wformat-security -std=c++17 -fopenmp -O3 -DNDEBUG -D_FORTIFY_SOURCE=2 -Wno-deprecated-declarations -Wno-attributes -msse4.1 -fstack-protector-strong -Wno-comment -Wno-subobject-linkage -I/home/gta/git/oneDNN/src  -Wmissing-field-initializers  -Wno-strict-overflow -Wno-maybe-uninitialized -DDNNL_ENABLE_ITT_TASKS -DDNNL_ENABLE_MAX_CPU_ISA -DDNNL_ENABLE_CPU_ISA_HINTS -I/home/gta/git/oneDNN/src/sycl" -fsycl -fvisibility-inlines-hidden  -ffp-model=precise -fno-reciprocal-math -Wno-unknown-warning-option -Wall -Wno-unknown-pragmas -fvisibility=internal -msse4.1  -fPIC -Wformat -Wformat-security -fstack-protector-all -Wno-unused-command-line-argument -fsycl-host-compiler=/usr/bin/g++   -Wno-pass-failed -Wno-tautological-compare -Wno-pass-failed -Wno-recommended-option -Wno-pass-failed  -O3 -DNDEBUG -D_FORTIFY_SOURCE=2 -MD -MT capi_stream.cpp.o -MF capi_stream.cpp.o.d -o capi_stream.cpp.o -c /home/gta/git/oneDNN/src/sycl/capi/capi_stream.cpp
```

--- a/cmake/host_compiler.cmake
+++ b/cmake/host_compiler.cmake
@@ -100,6 +100,11 @@ if(DPCPP_HOST_COMPILER_KIND MATCHES "^(GNU|CLANG)$")
     # Affects both, GNU and CLANG kinds.
     append(CMAKE_CXX_FLAGS "-Wno-unused-command-line-argument")
 
+    # Option `-fsycl-unnamed-lambda` is enabled by default, but not compatible
+    # with `-fsycl-host-compiler`. While icpx driver adds
+    # `-fno-sycl-unnamed-lambda` to avoid build issues clang++ does not do that.
+    append(CMAKE_CXX_FLAGS "-fno-sycl-unnamed-lambda")
+
     append(CMAKE_CXX_FLAGS "-fsycl-host-compiler=${DPCPP_HOST_COMPILER}")
     append_host_compiler_options(CMAKE_CXX_FLAGS "${DPCPP_HOST_COMPILER_OPTS}")
 endif()

